pkg/chains/storage/grafeas/grafeas_test.go

```diff
@@ -0,0 +1,382 @@
+						},
+					},
+				},
+			},
+		},
+	}
+	got, err := client.ListOccurrences(ctx,
+		&pb.ListOccurrencesRequest{
+			// This is just a placeholder.
+			// ProjectID doesn't matter here because we assume there is only one project in the mocked server.
+			Parent: fmt.Sprintf("project/PLACEHOLDER"),
+		},
+	)
+	if err != nil {
+		t.Fatal("Failed to create ListOccurrences. error ", err)
+	}
+
+	if !cmp.Equal(got, wanted, protocmp.Transform()) {
+		t.Errorf("Wrong list of occurrences received, got=%s", cmp.Diff(got, wanted, protocmp.Transform()))
+	}
+}
+
+// set up the connection between grafeas server and client
+// and return the client object to the caller
+func setupConnection() (*grpc.ClientConn, pb.GrafeasV1Beta1Client, error) {
+	serv := grpc.NewServer()
+	pb.RegisterGrafeasV1Beta1Server(serv, &grafeasServer)
+
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	go serv.Serve(lis)
+
+	conn, err := grpc.Dial(lis.Addr().String(), grpc.WithInsecure())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	client := pb.NewGrafeasV1Beta1Client(conn)
+	return conn, client, nil
+}
+
+// --------------------- Mocked GrafeasV1Beta1Server interface -----------------
+// https://pkg.go.dev/github.com/grafeas/grafeas@v0.2.0/proto/v1beta1/grafeas_go_proto#GrafeasV1Beta1Server
+func (s *mockGrafeasV1Beta1Server) CreateOccurrence(ctx context.Context, req *pb.CreateOccurrenceRequest) (*pb.Occurrence, error) {
+	if s.occurences == nil {
+		s.occurences = make(map[string]*pb.Occurrence)
+	}
+
+	occID := req.GetOccurrence().GetResource().GetUri()
+	expectedResponse := req.GetOccurrence()
+	expectedResponse.Name = occID // mock auto-generated id
+
+	s.occurences[occID] = expectedResponse
+	return expectedResponse, nil
+}
+
+func (s *mockGrafeasV1Beta1Server) CreateNote(ctx context.Context, req *pb.CreateNoteRequest) (*pb.Note, error) {
+	noteID := fmt.Sprintf("%s/notes/%s", req.GetParent(), req.GetNoteId())
+	expectedResponse := req.GetNote()
+	if s.notes == nil {
+		s.notes = make(map[string]*pb.Note)
+	}
+
+	if _, exists := s.notes[noteID]; exists {
+		return nil, gstatus.Error(codes.AlreadyExists, "note ID already exists")
+	}
+	s.notes[noteID] = expectedResponse
+	return expectedResponse, nil
+}
+
+func (s *mockGrafeasV1Beta1Server) ListOccurrences(ctx context.Context, req *pb.ListOccurrencesRequest) (*pb.ListOccurrencesResponse, error) {
+	occurrences := []*pb.Occurrence{}
+
+	if req.GetFilter() == "" {
+		// if filter is not specified, return all
+		for _, occ := range s.occurences {
+			occurrences = append(occurrences, occ)
+		}
+	} else {
+		// if filter is specified
+		// mock how filter works in ListOccurrencesRequest
+		uriFilter := strings.Split(req.GetFilter(), "=")[1] // url filter that has quotes
+		uriFilter = uriFilter[1 : len(uriFilter)-1]         // remove quotes
+
+		for id, occ := range s.occurences {
+			if id == uriFilter {
+				occurrences = append(occurrences, occ)
+			}
+		}
+	}
```

Comment:
You should document how exactly this can be used - if you're implementing non-obvious logic like filtering, you may want to consider adding additional tests. Alternatively, you can break the read tests into 2 separate components: 1) to make sure we are properly handling returned content 2) make sure we're constructing the right request.

Reply:
Great catch. Circular dependencies were introduced here. 
In the latest version, I added extra tests to make sure the mocked `ListOccurrences` function and its helper function of the filter logic can be trusted.